File: CPP Codes/nQueens.cpp
#include <bits/stdc++.h>
#define N 4
using namespace std;

// N Queens Problem Description:

/*
The N Queen is the problem of placing N chess queens on 
an N×N chessboard so that no two queens attack each other.
*/
/*Function to print solution */
void printBoard(int board[N][N])
{
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
            cout << " " << board[i][j] << " ";
        printf("\n");
    }
}


/*
A function to check whether it is safe to place the
queen on position (row,col) given the current configuration
of the board.
*/
bool safePos(int board[N][N], int row, int col)
{
    for (int i = 0; i < col; i++)
        if (board[row][i])
            return false;

    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
        if (board[i][j])
            return false;

    for (int i = row, j = col; j >= 0 && i < N; i++, j--)
        if (board[i][j])
            return false;

    return true;
}

/* A recursive utility function to solve N
   Queen problem */

// A recursive function to solve the n queens
// problem by placing a queen in each column
// one at a time
bool nQueensSolver(int board[N][N], int col)
{
    // queen placed in all columns
    if (col >= N)
    {
        return true;
    }
    // Checking all rows for column col
    for (int i = 0; i < N; i++)
    {

        if (safePos(board, i, col))
        {

            // Recursive step
            board[i][col] = 1;
            if (nQueensSolver(board, col + 1))
                return true;

            // Backtracking by undoing previous change
            board[i][col] = 0;
        }
    }

    return false;
}

// Main Driver Program
int main()
{
    int board[N][N] = {{0, 0, 0, 0},
                       {0, 0, 0, 0},
                       {0, 0, 0, 0},
                       {0, 0, 0, 0}};

    if (nQueensSolver(board, 0) == false)
    {
        cout << "nQueens Problem cannot be solved";
    }
    else
    {
        printBoard(board);
    }

    return 0;
}

// Time complexity: O(N!)
// Space Complexity: O(N^2)

File: Java Code/n_Queens.java
public class Q3{
    public static boolean issafe(char board[][] , int row, int col){
        // vertically up
        for(int i = row -1; i >= 0; i--){
            if(board[i][col] == 'Q'){
                return false;
            }
        }
        //diagonally left
        for(int i = row-1, j = col-1; i >= 0 && j>=0; i-- , j--){
            if(board[i][j] == 'Q'){
                return false;
            }
        }
        //diagonally right
        for(int i = row-1 , j = col +1; i >= 0 && j < board.length; i--,j++){
            if(board[i][j] == 'Q'){
                return false;
            }
        }
        return true;
    }
    public static boolean nQueens(char board[][], int row){
        // base case
        if(row == board.length){
            // printboard(board);
           
            return true;
        }
        // column loop 
        for(int j =0; j<board.length; j++){
            if(issafe(board, row , j)){
            board[row][j] = 'Q';
            if(nQueens(board, row+1)){
                return true;
            } // function call
            board[row][j] = '.'; // backtracking
            }
        }
        return false;
    }
    public static void printboard(char board[][]){
        System.out.println("----------------chess---------------");
        for(int i = 0; i<board.length; i++){
            for(int j = 0; j < board.length; j++){
                System.out.print(board[i][j] + " ");
            }
            System.out.println("");
        }
    }
    public static void main(String[] args) {
        int n =6;
        char board[][] = new char[n][n];
        // inttialize
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                board[i][j] = '.';
            }
        }
       if( nQueens(board,0)){
        System.out.println("Solution is possible");
        printboard(board);
       }
       else{
        System.out.println("Solution is not possible");
       }

    }
}
